### PR TITLE
Tactics inversion and replace work with eq in type

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** enhancement / fix

We do "minimal" changes so that `replace with` and `inversion` works with HoTT (on the side of HoTT, support for `discriminate` should also be provided, see HoTT/HoTT#1382).

This is a partial fix of #12846.

- [X] Added / updated test-suite (for `replace`)
- [ ] Corresponding documentation (that would be good that the role of each registerable constant is documented, but this goes beyond the scope of this PR)
- [X] Entry added in the changelog

### CODE REVIEW

**Comment by herbelin** [comment]
I realize that just for `core` registrations, there are currently 94 of them in the Coq archive (see below). There are otherwise 452 of them counting also registrations specific to plugins. So documenting everything would be difficult. The simplest would probably be to give to each of them the role of having to denote the exact CIC term that they currently have in the standard library. If we start giving a specific meaning to some items, like e.g. `core.eq.type` is any equivalence relation and not necessarily `eq`, that would lead us in a complex documentation to do.

Still, a registration for a default "equality" to use (independently of its exact CIC form) would be useful for tactics such as `inversion` (or `case_eq`).

Anyway, in the meantime, I'm afraid, @ppedrot, that you are the one in best position to assign for this PR.

```
core.False.type
core.IDProp.idProp
core.IDProp.type
core.JMeq.congr_canonical
core.JMeq.hom
core.JMeq.ind
core.JMeq.refl
core.JMeq.type
core.True.I
core.True.type
core.and.conj
core.and.type
core.ascii.ascii
core.ascii.type
core.bool.andb
core.bool.andb_prop
core.bool.andb_true_intro
core.bool.eqb
core.bool.false
core.bool.implb
core.bool.negb
core.bool.orb
core.bool.true
core.bool.type
core.bool.xorb
core.byte.type
core.comparison.Gt
core.comparison.type
core.dec.False
core.dec.True
core.dec.and
core.dec.dec_not_not
core.dec.iff
core.dec.imp
core.dec.imp_simp
core.dec.not
core.dec.not_and
core.dec.not_iff
core.dec.not_imp
core.dec.not_not
core.dec.not_or
core.dec.or
core.eq.congr
core.eq.congr2
core.eq.ind
core.eq.ind_r
core.eq.rect
core.eq.rect_r
core.eq.refl
core.eq.sind
core.eq.sind_r
core.eq.sym
core.eq.trans
core.eq.type
core.eq_true.type
core.eqdep_dec.inj_pair2
core.ex.type
core.identity.type
core.iff.proj1
core.iff.proj2
core.iff.type
core.list.cons
core.list.nil
core.list.type
core.nnpp.type
core.not.type
core.option.None
core.option.Some
core.option.type
core.or.type
core.prod.intro
core.prod.proj1
core.prod.proj2
core.prod.type
core.sig.intro
core.sig.proj1
core.sig.proj2
core.sig.type
core.sigT.existT
core.sigT.intro
core.sigT.proj1
core.sigT.proj2
core.sigT.type
core.string.empty
core.string.string
core.string.type
core.sum.inl
core.sum.inr
core.sumbool.type
core.unit.tt
core.unit.type
core.wf.acc
core.wf.acc_inv
core.wf.well_founded
```

File: tactics/inv.ml [hunk 1]
```diff
@@ -116,7 +116,11 @@ let make_inv_predicate env evd indf realargs id status concl =
   (* Now, we can recurse down this list, for each ai,(mkRel k) whether to
      push <Ai>(mkRel k)=ai (when   Ai is closed).
    In any case, we carry along the rest of pairs *)
-  let eqdata = Coqlib.build_coq_eq_data () in
+  let eqdata =
```
Comment (ppedrot) [inline tactics/inv.ml:119]: I think it's fragile to rely on the use of exceptions to check for the registration of this data, especially if we ever want to fully implement https://github.com/coq/coq/pull/11006. The most backwards compatible way I can think of is to export a mem function in Coqlib, but if you're willing to be slightly more disruptive I'd recommend returning an option there.

File: tactics/equality.ml [hunk 2]
```diff
@@ -659,8 +659,12 @@ let replace_using_leibniz clause c1 c2 l2r unsafe try_prove_eq_opt =
   | None ->
     tclFAIL 0 (str"Terms do not have convertible types")
   | Some evd ->
-    let e = lib_ref "core.eq.type" in
-    let sym = lib_ref "core.eq.sym" in
+    let e,sym =
+      try lib_ref "core.eq.type", lib_ref "core.eq.sym"
```
Comment (ppedrot) [inline tactics/equality.ml:663]: Same remark.

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
My comment is not a hard requirement for the PR but I think it'd be better to do it that way.

File: tactics/inv.ml (hunk 1 above)
Comment (herbelin) [reply tactics/inv.ml:119]: For 8.2.1, I would prefer not to be disruptive and to stay as it is, but, depending on @ejgallego's objectives wrt #11006, I can also prepare an alternative PR to #11006 which returns an option.

**Comment by ppedrot** [comment]
OK, will merge soon as-is then.

**Comment by Alizter** [comment]
Shall we also create an issue for documenting registerable constants?

**Comment by herbelin** [comment]
> Shall we also create an issue for documenting registerable constants?

We can try, if you're willing to help.

**Comment by ejgallego** [comment]
@ppedrot kind reminder to merge so we can ship in 8.12.1

> We can try, if you're willing to help.

Regading the issue about the constants it would be good to have a discussion on the semantic model, maybe for CUDW?

There were a lot of tricky things to account to when we did this change, so the current API is far from optimal, but we have improved quite a bit so IMHO we could try to understand what we'd like to ideally have.

File: tactics/inv.ml (hunk 1 above)
Comment (Zimmi48) [reply tactics/inv.ml:119]: It really seems that you always want to alpha-convert "8.12" into "8.2" @herbelin :laughing:

Comment (herbelin) [reply tactics/inv.ml:119]: It seems so indeed... Let me fix the comment.

**Pull request merged**